### ISSUE TITLE
IPv6 ping fails using sample application echo_server on FRDM-K64F

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1463]
**_Reported by ethan gao:_**

Build and flash the sample application of echo_server to board frdm-k64f, and then setup your client in another host with IP address 2001:0db8::2/64, ping the static IPv6 address a.k.a 2001:0db8::1 of the board, but no response received from the board. fortunately, the IPv4 ping works. maybe it's more or less related to the issue in GH-1410, and here to just track this issue on FRDM-K64F separately. 

(Imported from Jira ZEP-1602)

**Comment by nashif** [comment]
**_by ethan gao:_**

Just copy _Tomasz Bursztyka_  description here for reference, as follow:
---------------------------------------------------------------------------------------------------------------------------------------------------------

The mac address in frdm is randomly generated, so yes it will change at every boot.

I retried. Usually I setup my linux environment before starting the board: then it works.
If I do the other way round (i.e.: like you do, board start then linux setup). It does not.

So indeed there is something fishy here. On frdm that is. Can you create a ticket for it? This current ticket is for a101
------------------------------------------------------------------------------------------------------------------------------------------------------------

**Comment by nashif** [comment]
**_by ethan gao:_**

_Tomasz Bursztyka_  I tried your workaround to setup linux host firstly and then flash the echo_server sample to start frdm-k64f, but it doesn't work yet. My procedure is as follows:
Boot-up my linux host and then
    /* configure IPv4 address */
    $sudo ifconfig enp14s5 192.0.2.2
    /* configure IPv6 address */
    $sudo ifconfig enp14s5 add 2001:db8::2/64
    /* keep pinging the frdm-k64f IPv6 address */
   $ping6 -I enp14s5 2001:db8::1
Then
clone the zephyr repo and step into sample echo_server:
   run make pristine && make BOARD=frdm_k64f
   next copy the zephyr.bin to the mass USB folder to let it flash into frdm.
Finally find a NW cable to directly connect the linux host and the frdm-k64f.
unfortunately, the IPv6 ping fails as usual from the linux host, so I kept reseting the frdm, but the failure persists. BTW, the ping to IPv4 works.

So I want to make clear whether this procedure is your "I setup my linux environment before starting the board", or something missed by me ? please help point out if in that case, thank you !

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

I confirm this issue. So far I wasn't able to get ping6 the frdm_k64f despite various "before" vs "after" steps, so would like to join in asking Tomasz for exact steps to get it working.

What I see in Wiresharks is:

95	480.915031000	2001:db8::2	ff02::1:ff00:1	ICMPv6	86	Neighbor Solicitation for 2001:db8::1 from 3c:97:0e:c8:94:35

packets which never get answered. Nothing is visible as received in net console stats either (but fairly speaking, even IPv4 pings aren't counted as received).

As it's FRDM's MAC address is what can't be resolved, that no IPv6 networking works at all. All that stuff definitely works in QEMU.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

Actually, now IPv4 pings stopped working for me, board reset and power-offs don't help.

**Comment by nashif** [comment]
**_by Flavio Santes:_**

I confirm this issue exists for IPv6, ping6 is not working.

On the other hand, IPv4 works after a hard reset is applied and the Ethernet cable is unplugged (before the reset) and plugged again (after the reset).

Last, I don't observe this issue on TCP/IPv6 client-mode applications, so _perhaps_ this issue is related in some way to the server mode only.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

> Actually, now IPv4 pings stopped working for me, board reset and power-offs don't help

Over weekend, it cooled down and started working again, I didn't even have to reboot my laptop.

So, I proceeded with testing I wanted to do, and that's adding old good printf to eth_rx() in eth_mcux.c.

So, I do "ping 192.0.2.1", in Wireshark I see:


```
8	21.253887000	192.0.2.2	192.0.2.1	ICMP	98	Echo (ping) request  id=0x27d3, seq=1/256, ttl=64
```

And from board I get:
```
[dev/eth_mcux] [DBG] eth_rx: eth_rx: fr_len=98
```


Now I do "ping6 2001:db8::1", in Wireshark I see

```
195	337.188703000	2001:db8::2	ff02::1:ff00:1	ICMPv6	86	Neighbor Solicitation for 2001:db8::1 from 3c:97:0e:c8:94:35
```

And from the board I see - nothing, definitely no packet of size 86 gets received.

So, those pings, they go to multicast address ff02::1:ff00:1, and that needs to be added to multicast filters of the ethernet hardware, right? Is it being done properly?

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**


```
Interface 0x20000e20
====================
TX stack size 1200/1296 bytes unused 588 usage 612/1200 (46 %)
Link addr : 00:04:9F:B4:9E:B6
MTU       : 1500
IPv6 unicast addresses (max 5):
	fe80::204:9fff:feb4:9eb6 autoconf preferred infinite
	2001:db8::1 manual preferred infinite
IPv6 multicast addresses (max 5):
	ff84::2
IPv6 prefixes (max 2):
	<none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 17525
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
	192.0.2.1 manual preferred infinite
IPv4 multicast addresses (max 1):
	<none>
IPv4 gateway : 0.0.0.0
IPv4 netmask : 0.0.0.0
```

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

ENET_AddMulticastGroup() never called in eth_mcux.c.

Grepping for "cast" in zephyr/drivers/ethernet/ gives hits only for unicast and broadcast.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

> Last, I don't observe this issue on TCP/IPv6 client-mode applications, so perhaps this issue is related in some way to the server mode only.

Indeed, building echo_client (which lacks prj_frdm_k64f.conf) and running it, I can "ping6 2001:db8::1", but here's how it might work:


```
398	1998.658656000	2001:db8::1	ff02::1:ff00:2	ICMPv6	86	Neighbor Solicitation for 2001:db8::2 from 00:04:9f:93:2d:d6
399	1998.658686000	2001:db8::2	2001:db8::1	ICMPv6	86	Neighbor Advertisement 2001:db8::2 (sol, ovr) is at 3c:97:0e:c8:94:35
```

So, here FRDM actively sends Neighbor Solicitation, gets it reply, after which both FRDM and host are resolved to each other.

**Comment by nashif** [comment]
**_by Flavio Santes:_**

_Anas Nashif_ , _Mark Linkmeyer_ , _Andrei Laperie_  besides the common interest in having all the sample applications working as expected, it seems this issue is also affecting GH-727.

Is there any update about this?

Regards,
Flavio

**Comment by nashif** [comment]
**_by Anas Nashif:_**

this is high priority bug and must be fixed for 1.7, how does this affect http server sample exactly?

**Comment by nashif** [comment]
**_by Flavio Santes:_**

I observe the same behavior here than in my other TCP/IPv6-server-mode applications.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

I am not sure I am following, how does ipv6 ping related to http server sample? Can you please specify what is going wrong with the http server that is related to this issue?

**Comment by nashif** [comment]
**_by Flavio Santes:_**

_It seems_ that ND on the FRDM board is not working for TCP/IPv6 in server mode. ping6 errors are just a consequence of this. However, I would like to see Comms feedback just to be sure we refer to the same issue.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

> Actually, now IPv4 pings stopped working for me, board reset and power-offs don't help. 

Well, this is getting old. Just when I wanted to make a decisive test on this, I got this again. Actually, it wasn't just "no network communication to frdm_k64f", for few seconds my host completely froze as if under flood ping in old good days of Linux 1.2. So, apparently, frdm fights back. Plugging off eth cable fixed the freeze. For reference, e1000e driver is used on the host here.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

Few magic manipulations (which involve rmmod'ing and modprobe'ing e1000e - multiple times, one doesn't help, and attempts to suspend the box (which doesn't work - e1000e is a known suspend breaker here (x320/kernel 3.13.0-105-generic of ubuntu 14.04)), generally, it starts to work when you lose track of the steps and just try things at random) and ping4's are back.

And - 


```
# ping6  2001:db8::1
PING 2001:db8::1(2001:db8::1) 56 data bytes
64 bytes from 2001:db8::1: icmp_seq=1 ttl=64 time=1001 ms
64 bytes from 2001:db8::1: icmp_seq=2 ttl=64 time=1001 ms
64 bytes from 2001:db8::1: icmp_seq=3 ttl=64 time=1005 ms
```

What it took is:

```
+static uint8_t nd_mcast[6] = "\x33\x33\xff\x00\x00\x01";
+ENET_AddMulticastGroup(ENET, nd_mcast);
```


So, I propose to diagnose it as: *no* (see my comment above) Ethernet driver in Zephyr has support for multicast, which is a baseline feature for IPv6 support.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

(As an extra note, you can see funny ping times above. I get those slightly more often than "all eth comm is dead" situations described above. To prove that it can be non-funny too:


```
# ping6  2001:db8::1
PING 2001:db8::1(2001:db8::1) 56 data bytes
64 bytes from 2001:db8::1: icmp_seq=1 ttl=64 time=0.518 ms
64 bytes from 2001:db8::1: icmp_seq=2 ttl=64 time=0.323 ms
64 bytes from 2001:db8::1: icmp_seq=3 ttl=64 time=0.557 ms

```

)

**Comment by nashif** [comment]
**_by Jukka Rissanen:_**

I noticed this:

- start frdm
- ping zephyr ipv6 2001:db8::1 address from linux host
- linux sends NS (neighbor solicitation) but this packet is never received by zephyr (verified it by adding extensive logging in zephyr IP stack)
- if net-shell is enabled, then one can do "ping 2001:db8::2" from zephyr, this succeeds according to wireshark (linux receives NS and responds with NA)
- now one can ping 2001:db8::1 from linux as it already knows the mac address

This looks like identical case to what happened in enc28j60 ethernet driver which by default was discarding ipv6 multicast packets. See GH-1410 for details.

**Comment by nashif** [comment]
**_by Jukka Rissanen:_**

Assigning this to Maureen (mcux driver maintainer) as there is nothing that can be done in IP stack side for this issue.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

There're definitely things to be done on the IP stack side, as it doesn't (appears not to) properly integrate with hardware devices which handle multicasting on hardware level and such devices have (most definitely appear to have) problems with their driver model which doesn't allow such integration (in this respect the core of the problem may indeed lie in the driver model, but the IP stack is the leading stakeholder in that).

I'm finishing my research on the matter (which should handle "appear to"'s above) and will open a separate ticket on that.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

GH-1527 was submitted as a generalization of the issue discussed here and in GH-1410.

**Comment by nashif** [comment]
**_by Jing Wang:_**

Actually, in Qemu, there is 30% possibility "ping fail", see ND story at GH-714

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

_Jukka Rissanen_ , _Maureen Helm_ , _Paul Sokolovsky_ , who's the owner (aka assignee) for this bug?  It's a high priority bug which needs to be fixed in the 1.7 release.  It seems there might be some disagreement on who owns it.  I want to make sure the rightful owner is identified and agreed upon so it gets fixed.  Thx.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

I definitely agree, thanks for keeping an eye on such issues!

The workaround for 1.7.0 (ahead of better fix for all network drivers discussed in GH-1527, which definitely doesn't fit into 1.7) was submitted as https://gerrit.zephyrproject.org/r/10991 and merged into net branch. As an author of that patch, assigning ticket to myself. There's pending merge of net into master: https://gerrit.zephyrproject.org/r/11186 , after which this ticket can be re-verified and closed.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

(Status is "Code Review" because for me "Merged" means "Merged to master for everyone to benefit from". Maybe I'm wrong.)

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

The net branch merged into master.

**Comment by nashif** [comment]
**_by jiequan wu:_**

I tested the latest commit on master branch(commit id: 867e61cbf35d4a54fa2c). About IPV6 ping test, not at first, but it is fine after a while.
```
linux@linux-desktop:~/latest/zephyr.latest/samples/net/echo_server$ ping6 2001:0db8::1
PING 2001:0db8::1(2001:db8::1) 56 data bytes
64 bytes from 2001:db8::1: icmp_seq=1 ttl=64 time=0.386 ms
64 bytes from 2001:db8::1: icmp_seq=33 ttl=64 time=0.827 ms
64 bytes from 2001:db8::1: icmp_seq=34 ttl=64 time=0.332 ms
64 bytes from 2001:db8::1: icmp_seq=35 ttl=64 time=0.397 ms
64 bytes from 2001:db8::1: icmp_seq=36 ttl=64 time=0.433 ms
```
serial port output:
```
[echo-server] [INF] init_app: Run echo server
[echo-server] [INF] receive: Starting to wait
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=6
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=6
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=5
[dev/eth_mcux] [INF] eth_mcux_phy_event: Enabled 100M full-duplex mode.
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=6
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=6
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4
```

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

> About IPV6 ping test, not at first, but it is fine after a while.

I'm not sure what exactly you mean, but indeed, I saw cases (or maybe it's like that all the time) that it takes pings 1-2s to start go thru after a boot. I'd consider that "normal", there may be initialization ongoing during that time on both Zephyr IP stack and FRDM eth driver sides.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

> not at first, but it is fine after a while.

I actually saw that again in another ticket, so those are 3 independent reports, so I opened a separate GH-1587 on that.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

Besides the above (seperate) issue, I also retested this for master c99901560, works well.

**Comment by nashif** [comment]
**_by jiequan wu:_**

Tested the commit (e08d07c97d07235f0) on master branch, it works well.

**Comment by nashif** [comment]
**_by jiequan wu:_**

The latest commit on master branch(commit id: e4aa74125c), this issue reproduces. And on v1.7.0 branch is ok.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Setting FixVersion = 1.8 since it seems to be reoccurring on master, which means it needs to be fixed in 1.8 since it's a high priority bug.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

can tis still be reproduced? Please retest and report

**Comment by nashif** [comment]
**_by jiequan wu:_**

Yes, still tested failed. And IPv4 ping also failed by using the echo_server(commit id: eff52d9 && master branch).

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

Note that as of this moment, Ethernet of frdm_k64f is broken due to https://github.com/zephyrproject-rtos/zephyr/pull/261

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

Other patches required to unbreak this: https://github.com/zephyrproject-rtos/zephyr/pull/294, https://github.com/zephyrproject-rtos/zephyr/pull/295 . Please approve to get them sooner into master and v1.8-branch.

**Comment by nashif** [comment]
**_by jiequan wu:_**

Confirmed, it is fine(commit id: 596390 && master branch).

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

And just to be sure we can actually mark this as fixed for 1.8.0, I also retested against v1.8.0-rc2-21-g2b358a3ab from v1.8-branch. Both "ping6 2001:db8::1" and "telnet 2001:db8::1 4242" works immediately after rebooting frdm_k64f with disconnected ethernet cable and then connecting it and issuing each of the commands above.

**Comment by nashif** [comment]
**_by Maureen Helm:_**

Thanks _Paul Sokolovsky_  and _jiequan wu_ !

**Comment by nashif** [comment]
**_by jiequan wu:_**

Reproduce on v1.8.0-rc3-eff52d98542a.

**Comment by nashif** [comment]
**_by jiequan wu:_**

Confirmed, the latest (commit id:0b9f9f302374f8354b5fd0d736ad6c1151bdb677 && master branch) is fine. Maybe some patchs were missing on the v1.8.0-rc3 branch.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

From the last comment from _jiequan wu_  it appears this bug is present in the the RC3 of 1.8, right?  If that's the case, _Paul Sokolovsky_  are you working to fix it?  I ask because this is a high priority bug and in order to release 1.8 all high priority bugs need to be fixed.  Thanks.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

I cant reproduce this, works for both ipv4 and ipv6

**Comment by nashif** [comment]
**_by jiequan wu:_**

Reproduced on v1.8.0-rc3 branch and India team also found this.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

which commit on 1.8 branch are you testing?

**Comment by nashif** [comment]
**_by jiequan wu:_**

It is v1.8.0-rc3(commit id: eff52d98542ab826116841af95d6a3bc6057b2c5) , not v1.8-branch.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

This is not the right commit and does not match rc3, you should try to reproduce on the HEAd of the branch, ie.. 

8d963ce722e02ceb62b256b20b40a9a07e94779a

**Comment by nashif** [comment]
Blocks GH-727
